pages/content/amp-dev/documentation/guides-and-tutorials/develop/custom-javascript.md

```diff
@@ -0,0 +1,134 @@
+
+The `amp-script` component sends messages between the Web Worker thread and the main thread, causing any changes the user makes on the main DOM to be echoed on the Web Worker's false DOM. In turn, the Web Worker can then update the false DOM, which is reflected on the main DOM.
+
+[tip type="important"]
+ `<amp-script>` is in experimental mode and may break in unpredictable ways.
+[/tip]
+
+## Custom Scripts Caching 
+
+The [AMP cache](https://www.ampproject.org/docs/fundamentals/how_cached) serves custom JavaScript files inserted with `<amp-script>` same as AMP component scripts. This ensures that any custom JavaScript will not slow the speed of an AMP document. 
+
+The AMP cache proxies the JavaScript files and then delivers them. Users can expect the same performance experience from a page using `<amp-script>` as a page that doesn't include it. 
+
+# Limitations to `<amp-script>`
+
+To keep experiences on AMP pages consistent, limitations exist on `<amp-script>` to guarantee a fast loading time and smooth UI. 
+
+## Initialization
+
+JavaScript inside the Web Worker allows minimal change to the DOM on load. Changes allowed during this phase are:
+
+
+*   Registering event handlers.
+*   Splitting a TextNode into multiple TextNodes to allow for styling liberties, assuming the actual text doesn't change.
+
+The DOM inside `<amp-script>` tags should be almost identical before and after initialization. 
+
+
+## Interaction
+
+To ensure a responsive UX, DOM manipulations in response to a user interaction must happen in **less than one second**.  If a script mutates the DOM in response to a user interactions after one second, the `amp-script` component will terminate the Web Worker. A terminated `<amp-script>` component will not run again.
+
+
+## Script Size 
+
+AMP enforces a limit of 150 kilobytes of custom JavaScript on each page. This limit is shared  between every `<amp-script>` component on that page. If using a library, it must be imported to each individual `<amp-script>` element. 
+
+## Scope
+
+Any DOM elements the custom JavaScript files wishes to interact with must be wrapped inside the `<amp-script>` component tags, this includes other AMP components. The `<amp-script>` component considers `document.body` to be the `amp-script` element and not the document's `<body>` element.
+
+If you were to call `document.body.appendChild(document.createElement('span'))` within the `<amp-script>`file in the following document: 
+
+
+```html
+<body>  
+  <p>Hello!</p>
+  <div>
+    <amp-script layout="container" src="customjs.js">
+    </amp-script>
+  </div>
+</body>
+```
+
+
+It will result in this:
+
+
+```html
+<body>  
+  <p>Hello!</p>
+  <div>
+    <amp-script layout="container" src="customjs.js">
+      <span></span>
+    </amp-script>
+  </div>
+</body>
+```
+
+
+
+## Event Triggers 
```

Comment:
Remove this section, while this may change in the future, all events are currently allowed.